tests/lib/lazy_orm.py

```diff
@@ -0,0 +1,90 @@
+"""
+In order to write tests which can run using multiple different query engines we need
+fixtures which can be inserted into multiple different kinds of database. SQLAlchemy's
+ORM provides _almost_ enough abstraction to let us do this, but not quite. The sticking
+point is that we need to be able to customise the type of some columns on a per-database
+basis. For instance, when inserting into MSSQL we need `MSSQLDate()` and when inserting
+into Spark we need `SparkDate()`. SQLAlchemy itself provides no mechanism that I can see
+for doing this.
+
+The solution I have implemented below is a "lazy ORM" class. These looks pretty much
+like regular ORM classes when you define them, and their instances act just enough like
+ORM objects that you can build up a set a fixtures with them; but it's not until you
+"hydrate" these objects that they become actual ORM instances. This gives us the
+opportunity to customise the hydration process for each database and produce ORM classes
+with the appropriate column types.
+
+This is, obviously, a terrible thing. But it's at least terrible in a sort of generic
+Python way rather than reaching deep into the internals of SQLAlchemy. So I felt on
+balance that it was the least-worst of the many terrible solutions I tried.
+
+It also has the advantage of being reasonably self-contained and easy to back out: the
+lazy ORM classes are just a find-and-replace away from being regular ORM classes; and
+the hydration class is used in only a single place.
+"""
+import functools
+
+import sqlalchemy
+
+
+class LazyBase:
+    def __init_subclass__(cls, **kwargs):
+        # Keep a record of defined columns
+        cls._columns = {}
+        for key, value in vars(cls).items():
+            if isinstance(value, LazyColumn):
+                cls._columns[key] = value
+                # If we don't do this, instances of the class will have the column
+                # definition as their default column value, which is not what we want
+                setattr(cls, key, None)
+            elif isinstance(value, sqlalchemy.Column):
+                assert False, f"'{key}' must be a LazyColumn not a sqlalchemy.Column"
```

Comment:
Could this be an `assert not instance...` guard clause for the loop?

Reply:
Yes, that would be clearer, wouldn't it?